src/go/k8s/controllers/redpanda/cluster_controller_configuration.go

```diff
@@ -28,34 +28,35 @@ func (r *ClusterReconciler) markConfigurationChanged(
 		return false, nil
 	}
 
+	// We always need a last applied configuration annotation to use as baseline for centralized configuration patches
+	_, present, err := configMapResource.GetLastAppliedConfigurationFromCluster(ctx)
+	if err != nil {
+		return false, err
+	}
+	if !present {
+		// If no previous annotation was set for last applied configuration keys, then the ones in .boostrap.yaml are taken as reference
+		config, err := configMapResource.GetCurrentGlobalConfigurationFromCluster(ctx, configuration.DefaultCentralizedMode())
+		if err != nil {
+			return false, err
+		} else if config != nil {
+			if config.ClusterConfiguration == nil {
+				// Upgrade scenario: current cluster is not using centralized configuration.
+				// We need to extract bootstrap properties from the old format
```

Comment:
Bootstrap is only needed on first startup of a cluster: if a cluster is being upgraded from an earlier version, then you can skip it entirely (it's fine for the file to simply not exist).

Reply:
Right, the comment was not precise. The state indicates a specific upgrade from a version that was not using centralized configuration e.g. 21.11.1 to 22.1.1, so I think the bootstrap file will be used in this case by the new version.

I preferred anyway to always write the bootstrap file in the configmap, both to track the configuration more easily, but also to detect configuration drifts without having to call the admin API (which may not be available during some reconciliation loops).

I've changed this computation a bit in a [subsequent commit](https://github.com/redpanda-data/redpanda/pull/3978/commits/28a6ed92c761345bbf48f903a465152355411ecf) to simplify it.

Reply:
I should definitely change the term "drift", also in the code. The operator is not checking if the cluster is still configured as expected (as the term "drift" may imply), it's checking if the expected cluster configuration is different from the last-submitted one (stored in the configmap). Only after a "change" has been detected, the admin API is checked and any drift is fixed.

But e.g. if I set on the CR additional config `redpanda.segment_appender_flush_timeout_ms=2000` and someone changes that value manually to `3000` in the admin API, the operator will not set it back to `2000` until there's another change in the CR config section that triggers the merge process. We said that the users must be aware that they should not change the same configuration property in multiple places.

I did this verification on the CM instead of the cluster so that we're sure that, even in corner cases, the reconcile loop is not affected by the cluster state, i.e. it continues to work correctly when cluster is starting/restarting.

If detecting real drifts against the cluster is a desired feature, it's still possible to add a periodic check. Wdyt @RafalKorepta ?